**Building deep learning models with keras**

In this chapter, you'll use the keras library to build deep learning models for both regression as well as classification! You'll learn about the Specify-Compile-Fit workflow that you can use to make predictions and by the end of this chapter, you'll have all the tools necessary to build deep neural networks!

**Understanding your data**

You will soon start building models in Keras to predict wages based on various professional and demographic factors. Before you start building a model, it's good to understand your data by performing some exploratory analysis.

The data is pre-loaded into a pandas DataFrame called `df`. Use the `.head()` and `.describe()` methods in the IPython Shell for a quick overview of the DataFrame.

The target variable you'll be predicting is `wage_per_hour`. Some of the predictor variables are binary indicators, where a value of `1` represents `True`, and `0` represents `False`.

Of the `9` predictor variables in the DataFrame, how many are binary indicators? The min and max values as shown by `.describe()` will be informative here. How many binary indicator predictors are there?

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('hourly_wages.csv')

In [3]:
df.head()

wage_per_hour  union  education_yrs  experience_yrs  age  female  marr  \
0           5.10      0              8              21   35       1     1   
1           4.95      0              9              42   57       1     1   
2           6.67      0             12               1   19       0     0   
3           4.00      0             12               4   22       0     0   
4           7.50      0             12              17   35       0     1   

   south  manufacturing  construction  
0      0              1             0  
1      0              1             0  
2      0              1             0  
3      0              0             0  
4      0              0             0

In [4]:
df.tail()

wage_per_hour  union  education_yrs  experience_yrs  age  female  marr  \
529          11.36      0             18               5   29       0     0   
530           6.10      0             12              33   51       1     1   
531          23.25      1             17              25   48       1     1   
532          19.88      1             12              13   31       0     1   
533          15.38      0             16              33   55       0     1   

     south  manufacturing  construction  
529      0              0             0  
530      0              0             0  
531      0              0             0  
532      1              0             0  
533      0              1             0

In [5]:
df.describe()

wage_per_hour       union  education_yrs  experience_yrs         age  \
count     534.000000  534.000000     534.000000      534.000000  534.000000   
mean        9.024064    0.179775      13.018727       17.822097   36.833333   
std         5.139097    0.384360       2.615373       12.379710   11.726573   
min         1.000000    0.000000       2.000000        0.000000   18.000000   
25%         5.250000    0.000000      12.000000        8.000000   28.000000   
50%         7.780000    0.000000      12.000000       15.000000   35.000000   
75%        11.250000    0.000000      15.000000       26.000000   44.000000   
max        44.500000    1.000000      18.000000       55.000000   64.000000   

           female        marr       south  manufacturing  construction  
count  534.000000  534.000000  534.000000     534.000000    534.000000  
mean     0.458801    0.655431    0.292135       0.185393      0.044944  
std      0.498767    0.475673    0.455170       0.388981      0.207375  
min      0.000000    0.000000    0.000000       0.000000      0.000000  
25%      0.000000    0.000000    0.000000       0.000000      0.000000  
50%      0.000000    1.000000    0.000000       0.000000      0.000000  
75%      1.000000    1.000000    1.000000       0.000000      0.000000  
max      1.000000    1.000000    1.000000       1.000000      1.000000

**Specifying a model**

Now you'll get to work with your first model in Keras, and will immediately be able to run more complex neural network models on larger datasets compared to the first two chapters.

To start, you'll take the skeleton of a neural network and add a hidden layer and an output layer. You'll then fit that model and see Keras do the optimization so your model continually gets better.

As a start, you'll predict workers wages based on characteristics like their industry, education and level of experience. You can find the dataset in a pandas dataframe called `df`. For convenience, everything in `df` except for the target has been converted to a NumPy matrix called `predictors`. The target, `wage_per_hour`, is available as a NumPy matrix called `target`.

For all exercises in this chapter, we've imported the `Sequential` model constructor, the `Dense` layer constructor, and pandas.

In [6]:
predictors = df.iloc[:, 1:].values
predictors.shape

(534, 9)

In [7]:
target = df.iloc[:, 0].values
target.shape

(534,)

In [8]:
# Import necessary modules
import keras
from keras.layers import Dense
from keras.models import Sequential

# Save the number of columns in predictors: n_cols
n_cols = predictors.shape[1]

# Set up the model: model
model = Sequential()

# Add the first layer
model.add(Dense(50, activation='relu', input_shape=(n_cols,)))

# Add the second layer
model.add(Dense(32, activation='relu'))

# Add the output layer
model.add(Dense(1))


Using TensorFlow backend.


**Compiling the model**

You're now going to compile the model you specified earlier. To compile the model, you need to specify the optimizer and loss function to use. In the video, Dan mentioned that the Adam optimizer is an excellent choice. You can read more about it as well as other keras optimizers [here](https://keras.io/optimizers/#adam), and if you are really curious to learn more, you can read the [original paper](https://arxiv.org/abs/1412.6980v8) that introduced the Adam optimizer.

In this exercise, you'll use the Adam optimizer and the mean squared error loss function. Go for it!

In [9]:
# Import necessary modules
import keras
from keras.layers import Dense
from keras.models import Sequential

# Specify the model
n_cols = predictors.shape[1]
model = Sequential()
model.add(Dense(50, activation='relu', input_shape = (n_cols,)))
model.add(Dense(32, activation='relu'))
model.add(Dense(1))

# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error')

# Verify that model contains information from compiling
print("Loss function: " + model.loss)

Loss function: mean_squared_error


**Fitting the model**

You're at the most fun part. You'll now fit the model. Recall that the data to be used as predictive features is loaded in a NumPy matrix called `predictors` and the data to be predicted is stored in a NumPy matrix called target. Your model is pre-written and it has been compiled with the code from the previous exercise.

In [10]:
# Import necessary modules
import keras
from keras.layers import Dense
from keras.models import Sequential

# Specify the model
n_cols = predictors.shape[1]
model = Sequential()
model.add(Dense(50, activation='relu', input_shape = (n_cols,)))
model.add(Dense(32, activation='relu'))
model.add(Dense(1))

# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error')

# Fit the model
model.fit(predictors, target)


Epoch 1/10
534/534 [==============================] - 1s 1ms/step - loss: 112.1103
Epoch 2/10
534/534 [==============================] - 0s 58us/step - loss: 33.9149
Epoch 3/10
534/534 [==============================] - 0s 47us/step - loss: 29.9458
Epoch 4/10
534/534 [==============================] - 0s 52us/step - loss: 25.8971
Epoch 5/10
534/534 [==============================] - 0s 60us/step - loss: 23.7409
Epoch 6/10
534/534 [==============================] - 0s 51us/step - loss: 22.1118
Epoch 7/10
534/534 [==============================] - 0s 52us/step - loss: 21.7496
Epoch 8/10
534/534 [==============================] - 0s 51us/step - loss: 21.4748
Epoch 9/10
534/534 [==============================] - 0s 58us/step - loss: 21.3857
Epoch 10/10
534/534 [==============================] - 0s 49us/step - loss: 21.1268


Epoch 1/10

 32/534 [>.............................] - ETA: 0s - loss: 85.5178
534/534 [==============================] - 0s - loss: 45.9071     
Epoch 2/10

 32/534 [>.............................] - ETA: 0s - loss: 74.4328
534/534 [==============================] - 0s - loss: 28.9703     
Epoch 3/10

 32/534 [>.............................] - ETA: 0s - loss: 14.2559
534/534 [==============================] - 0s - loss: 23.4206     
Epoch 4/10

 32/534 [>.............................] - ETA: 0s - loss: 12.4489
534/534 [==============================] - 0s - loss: 21.4502     
Epoch 5/10

 32/534 [>.............................] - ETA: 0s - loss: 9.1276
534/534 [==============================] - 0s - loss: 21.4620    
Epoch 6/10

 32/534 [>.............................] - ETA: 0s - loss: 20.8471
534/534 [==============================] - 0s - loss: 21.3464     
Epoch 7/10

 32/534 [>.............................] - ETA: 0s - loss: 13.4813
534/534 [==============================] - 0s - loss: 21.1690     
Epoch 8/10

 32/534 [>.............................] - ETA: 0s - loss: 10.7564
534/534 [==============================] - 0s - loss: 21.1876     
Epoch 9/10

 32/534 [>.............................] - ETA: 0s - loss: 36.3705
534/534 [==============================] - 0s - loss: 21.6074     
Epoch 10/10

 32/534 [>.............................] - ETA: 0s - loss: 19.8555
534/534 [==============================] - 0s - loss: 20.9601
Out[1]: <keras.callbacks.History at 0x7f55b44cd5c0>

**Understanding your classification data**

Now you will start modeling with a new dataset for a classification problem. This data includes information about passengers on the Titanic. You will use predictors such as `age`, `fare` and where each passenger embarked from to predict who will survive. This data is from a [tutorial on data science competitions](https://www.kaggle.com/c/titanic). Look here for descriptions of the features.

The data is pre-loaded in a pandas DataFrame called `df`.

It's smart to review the maximum and minimum values of each variable to ensure the data isn't misformatted or corrupted. What was the maximum age of passengers on the Titanic? Use the `.describe()` method in the IPython Shell to answer this question.

In [11]:
df = pd.read_csv('titanic_all_numeric.csv')

In [12]:
df.head()

survived  pclass   age  sibsp  parch     fare  male age_was_missing  \
0         0       3  22.0      1      0   7.2500     1           False   
1         1       1  38.0      1      0  71.2833     0           False   
2         1       3  26.0      0      0   7.9250     0           False   
3         1       1  35.0      1      0  53.1000     0           False   
4         0       3  35.0      0      0   8.0500     1           False   

   embarked_from_cherbourg  embarked_from_queenstown  \
0                        0                         0   
1                        1                         0   
2                        0                         0   
3                        0                         0   
4                        0                         0   

   embarked_from_southampton  
0                          1  
1                          0  
2                          1  
3                          1  
4                          1

In [13]:
df.tail()

survived  pclass        age  sibsp  parch   fare  male age_was_missing  \
886         0       2  27.000000      0      0  13.00     1           False   
887         1       1  19.000000      0      0  30.00     0           False   
888         0       3  29.699118      1      2  23.45     0            True   
889         1       1  26.000000      0      0  30.00     1           False   
890         0       3  32.000000      0      0   7.75     1           False   

     embarked_from_cherbourg  embarked_from_queenstown  \
886                        0                         0   
887                        0                         0   
888                        0                         0   
889                        1                         0   
890                        0                         1   

     embarked_from_southampton  
886                          1  
887                          1  
888                          1  
889                          0  
890                          0

In [14]:
df.describe()

survived      pclass         age       sibsp       parch        fare  \
count  891.000000  891.000000  891.000000  891.000000  891.000000  891.000000   
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208   
std      0.486592    0.836071   13.002015    1.102743    0.806057   49.693429   
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000   
25%      0.000000    2.000000   22.000000    0.000000    0.000000    7.910400   
50%      0.000000    3.000000   29.699118    0.000000    0.000000   14.454200   
75%      1.000000    3.000000   35.000000    1.000000    0.000000   31.000000   
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200   

             male  embarked_from_cherbourg  embarked_from_queenstown  \
count  891.000000               891.000000                891.000000   
mean     0.647587                 0.188552                  0.086420   
std      0.477990                 0.391372                  0.281141   
min      0.000000                 0.000000                  0.000000   
25%      0.000000                 0.000000                  0.000000   
50%      1.000000                 0.000000                  0.000000   
75%      1.000000                 0.000000                  0.000000   
max      1.000000                 1.000000                  1.000000   

       embarked_from_southampton  
count                 891.000000  
mean                    0.722783  
std                     0.447876  
min                     0.000000  
25%                     0.000000  
50%                     1.000000  
75%                     1.000000  
max                     1.000000

**Last steps in classification models**

You'll now create a classification model using the titanic dataset, which has been pre-loaded into a DataFrame called `df`. You'll take information about the passengers and predict which ones survived.

The predictive variables are stored in a NumPy array `predictors`. The target to predict is in `df.survived`, though you'll have to manipulate it for keras. The number of predictive features is stored in `n_cols`.

Here, you'll use the `'sgd'` optimizer, which stands for [Stochastic Gradient Descent](https://en.wikipedia.org/wiki/Stochastic_gradient_descent). You'll learn more about this in the next chapter!

In [15]:
predictors = df.iloc[:, 1:].values
predictors.shape

(891, 10)

In [16]:
target = df.survived
target.shape

(891,)

In [17]:
n_cols = predictors.shape[1]
n_cols

10

{'Dense': keras.layers.core.Dense,
 'In': ['', 'vars()'],
 'Out': {},
 'Sequential': keras.models.Sequential,
 '_': '',
 '__': '',
 '___': '',
 '__builtin__': <module 'builtins' (built-in)>,
 '__builtins__': <module 'builtins' (built-in)>,
 '__name__': '__main__',
 '_dh': ['/tmp/tmpifkdt8fv'],
 '_i': '',
 '_i1': 'vars()',
 '_ih': ['', 'vars()'],
 '_ii': '',
 '_iii': '',
 '_oh': {},
 '_sh': <module 'IPython.core.shadowns' from '/var/lib/python/site-packages/IPython/core/shadowns.py'>,
 'df':      survived  pclass        age  sibsp  parch      fare  male  \
 0           0       3  22.000000      1      0    7.2500     1   
 1           1       1  38.000000      1      0   71.2833     0   
 2           1       3  26.000000      0      0    7.9250     0   
 3           1       1  35.000000      1      0   53.1000     0   
 4           0       3  35.000000      0      0    8.0500     1   
 5           0       3  29.699118      0      0    8.4583     1   
 6           0       1  54.000000      0      0   51.8625     1   
 7           0       3   2.000000      3      1   21.0750     1   
 8           1       3  27.000000      0      2   11.1333     0   
 9           1       2  14.000000      1      0   30.0708     0   
 10          1       3   4.000000      1      1   16.7000     0   
 11          1       1  58.000000      0      0   26.5500     0   
 12          0       3  20.000000      0      0    8.0500     1   
 13          0       3  39.000000      1      5   31.2750     1   
 14          0       3  14.000000      0      0    7.8542     0   
 15          1       2  55.000000      0      0   16.0000     0   
 16          0       3   2.000000      4      1   29.1250     1   
 17          1       2  29.699118      0      0   13.0000     1   
 18          0       3  31.000000      1      0   18.0000     0   
 19          1       3  29.699118      0      0    7.2250     0   
 20          0       2  35.000000      0      0   26.0000     1   
 21          1       2  34.000000      0      0   13.0000     1   
 22          1       3  15.000000      0      0    8.0292     0   
 23          1       1  28.000000      0      0   35.5000     1   
 24          0       3   8.000000      3      1   21.0750     0   
 25          1       3  38.000000      1      5   31.3875     0   
 26          0       3  29.699118      0      0    7.2250     1   
 27          0       1  19.000000      3      2  263.0000     1   
 28          1       3  29.699118      0      0    7.8792     0   
 29          0       3  29.699118      0      0    7.8958     1   
 ..        ...     ...        ...    ...    ...       ...   ...   
 861         0       2  21.000000      1      0   11.5000     1   
 862         1       1  48.000000      0      0   25.9292     0   
 863         0       3  29.699118      8      2   69.5500     0   
 864         0       2  24.000000      0      0   13.0000     1   
 865         1       2  42.000000      0      0   13.0000     0   
 866         1       2  27.000000      1      0   13.8583     0   
 867         0       1  31.000000      0      0   50.4958     1   
 868         0       3  29.699118      0      0    9.5000     1   
 869         1       3   4.000000      1      1   11.1333     1   
 870         0       3  26.000000      0      0    7.8958     1   
 871         1       1  47.000000      1      1   52.5542     0   
 872         0       1  33.000000      0      0    5.0000     1   
 873         0       3  47.000000      0      0    9.0000     1   
 874         1       2  28.000000      1      0   24.0000     0   
 875         1       3  15.000000      0      0    7.2250     0   
 876         0       3  20.000000      0      0    9.8458     1   
 877         0       3  19.000000      0      0    7.8958     1   
 878         0       3  29.699118      0      0    7.8958     1   
 879         1       1  56.000000      0      1   83.1583     0   
 880         1       2  25.000000      0      1   26.0000     0   
 881         0       3  33.000000      0      0    7.8958     1   
 882         0       3  22.000000      0      0   10.5167     0   
 883         0       2  28.000000      0      0   10.5000     1   
 884         0       3  25.000000      0      0    7.0500     1   
 885         0       3  39.000000      0      5   29.1250     0   
 886         0       2  27.000000      0      0   13.0000     1   
 887         1       1  19.000000      0      0   30.0000     0   
 888         0       3  29.699118      1      2   23.4500     0   
 889         1       1  26.000000      0      0   30.0000     1   
 890         0       3  32.000000      0      0    7.7500     1   
 
      age_was_missing  embarked_from_cherbourg  embarked_from_queenstown  \
 0                  0                        0                         0   
 1                  0                        1                         0   
 2                  0                        0                         0   
 3                  0                        0                         0   
 4                  0                        0                         0   
 5                  1                        0                         1   
 6                  0                        0                         0   
 7                  0                        0                         0   
 8                  0                        0                         0   
 9                  0                        1                         0   
 10                 0                        0                         0   
 11                 0                        0                         0   
 12                 0                        0                         0   
 13                 0                        0                         0   
 14                 0                        0                         0   
 15                 0                        0                         0   
 16                 0                        0                         1   
 17                 1                        0                         0   
 18                 0                        0                         0   
 19                 1                        1                         0   
 20                 0                        0                         0   
 21                 0                        0                         0   
 22                 0                        0                         1   
 23                 0                        0                         0   
 24                 0                        0                         0   
 25                 0                        0                         0   
 26                 1                        1                         0   
 27                 0                        0                         0   
 28                 1                        0                         1   
 29                 1                        0                         0   
 ..               ...                      ...                       ...   
 861                0                        0                         0   
 862                0                        0                         0   
 863                1                        0                         0   
 864                0                        0                         0   
 865                0                        0                         0   
 866                0                        1                         0   
 867                0                        0                         0   
 868                1                        0                         0   
 869                0                        0                         0   
 870                0                        0                         0   
 871                0                        0                         0   
 872                0                        0                         0   
 873                0                        0                         0   
 874                0                        1                         0   
 875                0                        1                         0   
 876                0                        0                         0   
 877                0                        0                         0   
 878                1                        0                         0   
 879                0                        1                         0   
 880                0                        0                         0   
 881                0                        0                         0   
 882                0                        0                         0   
 883                0                        0                         0   
 884                0                        0                         0   
 885                0                        0                         1   
 886                0                        0                         0   
 887                0                        0                         0   
 888                1                        0                         0   
 889                0                        1                         0   
 890                0                        0                         1   
 
      embarked_from_southampton  
 0                            1  
 1                            0  
 2                            1  
 3                            1  
 4                            1  
 5                            0  
 6                            1  
 7                            1  
 8                            1  
 9                            0  
 10                           1  
 11                           1  
 12                           1  
 13                           1  
 14                           1  
 15                           1  
 16                           0  
 17                           1  
 18                           1  
 19                           0  
 20                           1  
 21                           1  
 22                           0  
 23                           1  
 24                           1  
 25                           1  
 26                           0  
 27                           1  
 28                           0  
 29                           1  
 ..                         ...  
 861                          1  
 862                          1  
 863                          1  
 864                          1  
 865                          1  
 866                          0  
 867                          1  
 868                          1  
 869                          1  
 870                          1  
 871                          1  
 872                          1  
 873                          1  
 874                          0  
 875                          0  
 876                          1  
 877                          1  
 878                          1  
 879                          0  
 880                          1  
 881                          1  
 882                          1  
 883                          1  
 884                          1  
 885                          0  
 886                          1  
 887                          1  
 888                          1  
 889                          0  
 890                          0  
 
 [891 rows x 11 columns],
 'exit': <IPython.core.autocall.ExitAutocall at 0x7f55c3a36400>,
 'get_ipython': <bound method InteractiveShell.get_ipython of <IPython.core.interactiveshell.InteractiveShell object at 0x7f55b8453be0>>,
 'keras': <module 'keras' from '/usr/local/lib/python3.5/dist-packages/keras/__init__.py'>,
 'n_cols': 10,
 'pd': <module 'pandas' from '/var/lib/python/site-packages/pandas/__init__.py'>,
 'predictors': array([[  3.        ,  22.        ,   1.        , ...,   0.        ,
           0.        ,   1.        ],
        [  1.        ,  38.        ,   1.        , ...,   1.        ,
           0.        ,   0.        ],
        [  3.        ,  26.        ,   0.        , ...,   0.        ,
           0.        ,   1.        ],
        ..., 
        [  3.        ,  29.69911765,   1.        , ...,   0.        ,
           0.        ,   1.        ],
        [  1.        ,  26.        ,   0.        , ...,   1.        ,
           0.        ,   0.        ],
        [  3.        ,  32.        ,   0.        , ...,   0.        ,
           1.        ,   0.        ]]),
 'quit': <IPython.core.autocall.ExitAutocall at 0x7f55c3a36400>,
 'to_categorical': <function keras.utils.np_utils.to_categorical>}

In [18]:
# Import necessary modules
import keras
from keras.layers import Dense
from keras.models import Sequential
from keras.utils import to_categorical

# Convert the target to categorical: target
target = to_categorical(df.survived)

# Set up the model
model = Sequential()

# Add the first layer
model.add(Dense(32, activation='relu', input_shape = (n_cols,)))

# Add the output layer
model.add(Dense(2, activation='softmax'))  

# Compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

# Fit the model
model.fit(predictors, target)


Epoch 1/10
891/891 [==============================] - 0s 374us/step - loss: 3.6592 - acc: 0.5645
Epoch 2/10
891/891 [==============================] - 0s 44us/step - loss: 2.0811 - acc: 0.6038
Epoch 3/10
891/891 [==============================] - 0s 43us/step - loss: 1.0067 - acc: 0.6364
Epoch 4/10
891/891 [==============================] - 0s 46us/step - loss: 0.7448 - acc: 0.6588
Epoch 5/10
891/891 [==============================] - 0s 45us/step - loss: 0.6774 - acc: 0.6824
Epoch 6/10
891/891 [==============================] - 0s 45us/step - loss: 0.6267 - acc: 0.6891
Epoch 7/10
891/891 [==============================] - ETA: 0s - loss: 0.7115 - acc: 0.656 - 0s 38us/step - loss: 0.6355 - acc: 0.6700
Epoch 8/10
891/891 [==============================] - 0s 39us/step - loss: 0.5934 - acc: 0.6869
Epoch 9/10
891/891 [==============================] - 0s 45us/step - loss: 0.5968 - acc: 0.6869
Epoch 10/10
891/891 [==============================] - 0s 42us/step - loss: 0.5870 - acc: 0.7015


Epoch 1/10

 32/891 [>.............................] - ETA: 0s - loss: 2.2860 - acc: 0.4688
480/891 [===============>..............] - ETA: 0s - loss: 2.7017 - acc: 0.5771
891/891 [==============================] - 0s - loss: 2.0223 - acc: 0.6150     
Epoch 2/10

 32/891 [>.............................] - ETA: 0s - loss: 1.3087 - acc: 0.4062
480/891 [===============>..............] - ETA: 0s - loss: 1.1448 - acc: 0.6229
864/891 [============================>.] - ETA: 0s - loss: 1.1202 - acc: 0.6169
891/891 [==============================] - 0s - loss: 1.1082 - acc: 0.6184     
Epoch 3/10

 32/891 [>.............................] - ETA: 0s - loss: 1.2949 - acc: 0.7500
416/891 [=============>................] - ETA: 0s - loss: 0.9214 - acc: 0.6154
800/891 [=========================>....] - ETA: 0s - loss: 0.9359 - acc: 0.6138
891/891 [==============================] - 0s - loss: 0.9109 - acc: 0.6173     
Epoch 4/10

 32/891 [>.............................] - ETA: 0s - loss: 0.8078 - acc: 0.5625
576/891 [==================>...........] - ETA: 0s - loss: 0.7991 - acc: 0.6597
891/891 [==============================] - 0s - loss: 0.7541 - acc: 0.6566     
Epoch 5/10

 32/891 [>.............................] - ETA: 0s - loss: 0.7259 - acc: 0.6562
704/891 [======================>.......] - ETA: 0s - loss: 0.7659 - acc: 0.6605
891/891 [==============================] - 0s - loss: 0.7704 - acc: 0.6465     
Epoch 6/10

 32/891 [>.............................] - ETA: 0s - loss: 0.5819 - acc: 0.5938
448/891 [==============>...............] - ETA: 0s - loss: 0.7569 - acc: 0.6540
891/891 [==============================] - 0s - loss: 0.7122 - acc: 0.6532     
Epoch 7/10

 32/891 [>.............................] - ETA: 0s - loss: 0.7170 - acc: 0.6875
480/891 [===============>..............] - ETA: 0s - loss: 0.6419 - acc: 0.6708
891/891 [==============================] - 0s - loss: 0.6810 - acc: 0.6813     
Epoch 8/10

 32/891 [>.............................] - ETA: 0s - loss: 0.5142 - acc: 0.7812
576/891 [==================>...........] - ETA: 0s - loss: 0.6646 - acc: 0.6858
891/891 [==============================] - 0s - loss: 0.6461 - acc: 0.6790     
Epoch 9/10

 32/891 [>.............................] - ETA: 0s - loss: 0.7385 - acc: 0.6562
416/891 [=============>................] - ETA: 0s - loss: 0.6555 - acc: 0.6587
800/891 [=========================>....] - ETA: 0s - loss: 0.6653 - acc: 0.6675
891/891 [==============================] - 0s - loss: 0.6660 - acc: 0.6723     
Epoch 10/10

 32/891 [>.............................] - ETA: 0s - loss: 0.9764 - acc: 0.5938
352/891 [==========>...................] - ETA: 0s - loss: 0.6541 - acc: 0.6250
672/891 [=====================>........] - ETA: 0s - loss: 0.6135 - acc: 0.6771
891/891 [==============================] - 0s - loss: 0.6132 - acc: 0.6745
Out[2]: <keras.callbacks.History at 0x7f55b4c83fd0>

**Making predictions**

The trained network from your previous coding exercise is now stored as `model`. New data to make predictions is stored in a NumPy array as `pred_data`. Use `model` to make predictions on your new data.

In this exercise, your predictions will be probabilities, which is the most common way for data scientists to communicate their predictions to colleagues.

In [19]:
pred_data = np.array([[2, 34.0, 0, 0, 13.0, 1, False, 0, 0, 1],
        [2, 31.0, 1, 1, 26.25, 0, False, 0, 0, 1],
        [1, 11.0, 1, 2, 120.0, 1, False, 0, 0, 1],
        [3, 0.42, 0, 1, 8.5167, 1, False, 1, 0, 0],
        [3, 27.0, 0, 0, 6.975, 1, False, 0, 0, 1],
        [3, 31.0, 0, 0, 7.775, 1, False, 0, 0, 1],
        [1, 39.0, 0, 0, 0.0, 1, False, 0, 0, 1],
        [3, 18.0, 0, 0, 7.775, 0, False, 0, 0, 1],
        [2, 39.0, 0, 0, 13.0, 1, False, 0, 0, 1],
        [1, 33.0, 1, 0, 53.1, 0, False, 0, 0, 1],
        [3, 26.0, 0, 0, 7.8875, 1, False, 0, 0, 1],
        [3, 39.0, 0, 0, 24.15, 1, False, 0, 0, 1],
        [2, 35.0, 0, 0, 10.5, 1, False, 0, 0, 1],
        [3, 6.0, 4, 2, 31.275, 0, False, 0, 0, 1],
        [3, 30.5, 0, 0, 8.05, 1, False, 0, 0, 1],
        [1, 29.69911764705882, 0, 0, 0.0, 1, True, 0, 0, 1],
        [3, 23.0, 0, 0, 7.925, 0, False, 0, 0, 1],
        [2, 31.0, 1, 1, 37.0042, 1, False, 1, 0, 0],
        [3, 43.0, 0, 0, 6.45, 1, False, 0, 0, 1],
        [3, 10.0, 3, 2, 27.9, 1, False, 0, 0, 1],
        [1, 52.0, 1, 1, 93.5, 0, False, 0, 0, 1],
        [3, 27.0, 0, 0, 8.6625, 1, False, 0, 0, 1],
        [1, 38.0, 0, 0, 0.0, 1, False, 0, 0, 1],
        [3, 27.0, 0, 1, 12.475, 0, False, 0, 0, 1],
        [3, 2.0, 4, 1, 39.6875, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 0, 0, 6.95, 1, True, 0, 1, 0],
        [3, 29.69911764705882, 0, 0, 56.4958, 1, True, 0, 0, 1],
        [2, 1.0, 0, 2, 37.0042, 1, False, 1, 0, 0],
        [3, 29.69911764705882, 0, 0, 7.75, 1, True, 0, 1, 0],
        [1, 62.0, 0, 0, 80.0, 0, False, 0, 0, 0],
        [3, 15.0, 1, 0, 14.4542, 0, False, 1, 0, 0],
        [2, 0.83, 1, 1, 18.75, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 0, 0, 7.2292, 1, True, 1, 0, 0],
        [3, 23.0, 0, 0, 7.8542, 1, False, 0, 0, 1],
        [3, 18.0, 0, 0, 8.3, 1, False, 0, 0, 1],
        [1, 39.0, 1, 1, 83.1583, 0, False, 1, 0, 0],
        [3, 21.0, 0, 0, 8.6625, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 0, 0, 8.05, 1, True, 0, 0, 1],
        [3, 32.0, 0, 0, 56.4958, 1, False, 0, 0, 1],
        [1, 29.69911764705882, 0, 0, 29.7, 1, True, 1, 0, 0],
        [3, 20.0, 0, 0, 7.925, 1, False, 0, 0, 1],
        [2, 16.0, 0, 0, 10.5, 1, False, 0, 0, 1],
        [1, 30.0, 0, 0, 31.0, 0, False, 1, 0, 0],
        [3, 34.5, 0, 0, 6.4375, 1, False, 1, 0, 0],
        [3, 17.0, 0, 0, 8.6625, 1, False, 0, 0, 1],
        [3, 42.0, 0, 0, 7.55, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 8, 2, 69.55, 1, True, 0, 0, 1],
        [3, 35.0, 0, 0, 7.8958, 1, False, 1, 0, 0],
        [2, 28.0, 0, 1, 33.0, 1, False, 0, 0, 1],
        [1, 29.69911764705882, 1, 0, 89.1042, 0, True, 1, 0, 0],
        [3, 4.0, 4, 2, 31.275, 1, False, 0, 0, 1],
        [3, 74.0, 0, 0, 7.775, 1, False, 0, 0, 1],
        [3, 9.0, 1, 1, 15.2458, 0, False, 1, 0, 0],
        [1, 16.0, 0, 1, 39.4, 0, False, 0, 0, 1],
        [2, 44.0, 1, 0, 26.0, 0, False, 0, 0, 1],
        [3, 18.0, 0, 1, 9.35, 0, False, 0, 0, 1],
        [1, 45.0, 1, 1, 164.8667, 0, False, 0, 0, 1],
        [1, 51.0, 0, 0, 26.55, 1, False, 0, 0, 1],
        [3, 24.0, 0, 3, 19.2583, 0, False, 1, 0, 0],
        [3, 29.69911764705882, 0, 0, 7.2292, 1, True, 1, 0, 0],
        [3, 41.0, 2, 0, 14.1083, 1, False, 0, 0, 1],
        [2, 21.0, 1, 0, 11.5, 1, False, 0, 0, 1],
        [1, 48.0, 0, 0, 25.9292, 0, False, 0, 0, 1],
        [3, 29.69911764705882, 8, 2, 69.55, 0, True, 0, 0, 1],
        [2, 24.0, 0, 0, 13.0, 1, False, 0, 0, 1],
        [2, 42.0, 0, 0, 13.0, 0, False, 0, 0, 1],
        [2, 27.0, 1, 0, 13.8583, 0, False, 1, 0, 0],
        [1, 31.0, 0, 0, 50.4958, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 0, 0, 9.5, 1, True, 0, 0, 1],
        [3, 4.0, 1, 1, 11.1333, 1, False, 0, 0, 1],
        [3, 26.0, 0, 0, 7.8958, 1, False, 0, 0, 1],
        [1, 47.0, 1, 1, 52.5542, 0, False, 0, 0, 1],
        [1, 33.0, 0, 0, 5.0, 1, False, 0, 0, 1],
        [3, 47.0, 0, 0, 9.0, 1, False, 0, 0, 1],
        [2, 28.0, 1, 0, 24.0, 0, False, 1, 0, 0],
        [3, 15.0, 0, 0, 7.225, 0, False, 1, 0, 0],
        [3, 20.0, 0, 0, 9.8458, 1, False, 0, 0, 1],
        [3, 19.0, 0, 0, 7.8958, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 0, 0, 7.8958, 1, True, 0, 0, 1],
        [1, 56.0, 0, 1, 83.1583, 0, False, 1, 0, 0],
        [2, 25.0, 0, 1, 26.0, 0, False, 0, 0, 1],
        [3, 33.0, 0, 0, 7.8958, 1, False, 0, 0, 1],
        [3, 22.0, 0, 0, 10.5167, 0, False, 0, 0, 1],
        [2, 28.0, 0, 0, 10.5, 1, False, 0, 0, 1],
        [3, 25.0, 0, 0, 7.05, 1, False, 0, 0, 1],
        [3, 39.0, 0, 5, 29.125, 0, False, 0, 1, 0],
        [2, 27.0, 0, 0, 13.0, 1, False, 0, 0, 1],
        [1, 19.0, 0, 0, 30.0, 0, False, 0, 0, 1],
        [3, 29.69911764705882, 1, 2, 23.45, 0, True, 0, 0, 1],
        [1, 26.0, 0, 0, 30.0, 1, False, 1, 0, 0],
        [3, 32.0, 0, 0, 7.75, 1, False, 0, 1, 0]])
train_size = 800

{'Dense': keras.layers.core.Dense,
 'In': ['', 'vars()'],
 'Out': {},
 'Sequential': keras.models.Sequential,
 'X': array([[3, 22.0, 1, ..., 0, 0, 1],
        [1, 38.0, 1, ..., 1, 0, 0],
        [3, 26.0, 0, ..., 0, 0, 1],
        ..., 
        [3, 29.69911764705882, 1, ..., 0, 0, 1],
        [1, 26.0, 0, ..., 1, 0, 0],
        [3, 32.0, 0, ..., 0, 1, 0]], dtype=object),
 '_': '',
 '__': '',
 '___': '',
 '__builtin__': <module 'builtins' (built-in)>,
 '__builtins__': <module 'builtins' (built-in)>,
 '__name__': '__main__',
 '_dh': ['/tmp/tmpd9pr98uj'],
 '_i': '',
 '_i1': 'vars()',
 '_ih': ['', 'vars()'],
 '_ii': '',
 '_iii': '',
 '_oh': {},
 '_sh': <module 'IPython.core.shadowns' from '/var/lib/python/site-packages/IPython/core/shadowns.py'>,
 'data':      survived  pclass        age  sibsp  parch      fare  male  \
 0           0       3  22.000000      1      0    7.2500     1   
 1           1       1  38.000000      1      0   71.2833     0   
 2           1       3  26.000000      0      0    7.9250     0   
 3           1       1  35.000000      1      0   53.1000     0   
 4           0       3  35.000000      0      0    8.0500     1   
 5           0       3  29.699118      0      0    8.4583     1   
 6           0       1  54.000000      0      0   51.8625     1   
 7           0       3   2.000000      3      1   21.0750     1   
 8           1       3  27.000000      0      2   11.1333     0   
 9           1       2  14.000000      1      0   30.0708     0   
 10          1       3   4.000000      1      1   16.7000     0   
 11          1       1  58.000000      0      0   26.5500     0   
 12          0       3  20.000000      0      0    8.0500     1   
 13          0       3  39.000000      1      5   31.2750     1   
 14          0       3  14.000000      0      0    7.8542     0   
 15          1       2  55.000000      0      0   16.0000     0   
 16          0       3   2.000000      4      1   29.1250     1   
 17          1       2  29.699118      0      0   13.0000     1   
 18          0       3  31.000000      1      0   18.0000     0   
 19          1       3  29.699118      0      0    7.2250     0   
 20          0       2  35.000000      0      0   26.0000     1   
 21          1       2  34.000000      0      0   13.0000     1   
 22          1       3  15.000000      0      0    8.0292     0   
 23          1       1  28.000000      0      0   35.5000     1   
 24          0       3   8.000000      3      1   21.0750     0   
 25          1       3  38.000000      1      5   31.3875     0   
 26          0       3  29.699118      0      0    7.2250     1   
 27          0       1  19.000000      3      2  263.0000     1   
 28          1       3  29.699118      0      0    7.8792     0   
 29          0       3  29.699118      0      0    7.8958     1   
 ..        ...     ...        ...    ...    ...       ...   ...   
 861         0       2  21.000000      1      0   11.5000     1   
 862         1       1  48.000000      0      0   25.9292     0   
 863         0       3  29.699118      8      2   69.5500     0   
 864         0       2  24.000000      0      0   13.0000     1   
 865         1       2  42.000000      0      0   13.0000     0   
 866         1       2  27.000000      1      0   13.8583     0   
 867         0       1  31.000000      0      0   50.4958     1   
 868         0       3  29.699118      0      0    9.5000     1   
 869         1       3   4.000000      1      1   11.1333     1   
 870         0       3  26.000000      0      0    7.8958     1   
 871         1       1  47.000000      1      1   52.5542     0   
 872         0       1  33.000000      0      0    5.0000     1   
 873         0       3  47.000000      0      0    9.0000     1   
 874         1       2  28.000000      1      0   24.0000     0   
 875         1       3  15.000000      0      0    7.2250     0   
 876         0       3  20.000000      0      0    9.8458     1   
 877         0       3  19.000000      0      0    7.8958     1   
 878         0       3  29.699118      0      0    7.8958     1   
 879         1       1  56.000000      0      1   83.1583     0   
 880         1       2  25.000000      0      1   26.0000     0   
 881         0       3  33.000000      0      0    7.8958     1   
 882         0       3  22.000000      0      0   10.5167     0   
 883         0       2  28.000000      0      0   10.5000     1   
 884         0       3  25.000000      0      0    7.0500     1   
 885         0       3  39.000000      0      5   29.1250     0   
 886         0       2  27.000000      0      0   13.0000     1   
 887         1       1  19.000000      0      0   30.0000     0   
 888         0       3  29.699118      1      2   23.4500     0   
 889         1       1  26.000000      0      0   30.0000     1   
 890         0       3  32.000000      0      0    7.7500     1   
 
     age_was_missing  embarked_from_cherbourg  embarked_from_queenstown  \
 0             False                        0                         0   
 1             False                        1                         0   
 2             False                        0                         0   
 3             False                        0                         0   
 4             False                        0                         0   
 5              True                        0                         1   
 6             False                        0                         0   
 7             False                        0                         0   
 8             False                        0                         0   
 9             False                        1                         0   
 10            False                        0                         0   
 11            False                        0                         0   
 12            False                        0                         0   
 13            False                        0                         0   
 14            False                        0                         0   
 15            False                        0                         0   
 16            False                        0                         1   
 17             True                        0                         0   
 18            False                        0                         0   
 19             True                        1                         0   
 20            False                        0                         0   
 21            False                        0                         0   
 22            False                        0                         1   
 23            False                        0                         0   
 24            False                        0                         0   
 25            False                        0                         0   
 26             True                        1                         0   
 27            False                        0                         0   
 28             True                        0                         1   
 29             True                        0                         0   
 ..              ...                      ...                       ...   
 861           False                        0                         0   
 862           False                        0                         0   
 863            True                        0                         0   
 864           False                        0                         0   
 865           False                        0                         0   
 866           False                        1                         0   
 867           False                        0                         0   
 868            True                        0                         0   
 869           False                        0                         0   
 870           False                        0                         0   
 871           False                        0                         0   
 872           False                        0                         0   
 873           False                        0                         0   
 874           False                        1                         0   
 875           False                        1                         0   
 876           False                        0                         0   
 877           False                        0                         0   
 878            True                        0                         0   
 879           False                        1                         0   
 880           False                        0                         0   
 881           False                        0                         0   
 882           False                        0                         0   
 883           False                        0                         0   
 884           False                        0                         0   
 885           False                        0                         1   
 886           False                        0                         0   
 887           False                        0                         0   
 888            True                        0                         0   
 889           False                        1                         0   
 890           False                        0                         1   
 
      embarked_from_southampton  
 0                            1  
 1                            0  
 2                            1  
 3                            1  
 4                            1  
 5                            0  
 6                            1  
 7                            1  
 8                            1  
 9                            0  
 10                           1  
 11                           1  
 12                           1  
 13                           1  
 14                           1  
 15                           1  
 16                           0  
 17                           1  
 18                           1  
 19                           0  
 20                           1  
 21                           1  
 22                           0  
 23                           1  
 24                           1  
 25                           1  
 26                           0  
 27                           1  
 28                           0  
 29                           1  
 ..                         ...  
 861                          1  
 862                          1  
 863                          1  
 864                          1  
 865                          1  
 866                          0  
 867                          1  
 868                          1  
 869                          1  
 870                          1  
 871                          1  
 872                          1  
 873                          1  
 874                          0  
 875                          0  
 876                          1  
 877                          1  
 878                          1  
 879                          0  
 880                          1  
 881                          1  
 882                          1  
 883                          1  
 884                          1  
 885                          0  
 886                          1  
 887                          1  
 888                          1  
 889                          0  
 890                          0  
 
 [891 rows x 11 columns],
 'exit': <IPython.core.autocall.ExitAutocall at 0x7f55b790e400>,
 'get_ipython': <bound method InteractiveShell.get_ipython of <IPython.core.interactiveshell.InteractiveShell object at 0x7f55b8453be0>>,
 'keras': <module 'keras' from '/usr/local/lib/python3.5/dist-packages/keras/__init__.py'>,
 'n_cols': 10,
 'pd': <module 'pandas' from '/var/lib/python/site-packages/pandas/__init__.py'>,
 'pred_data': array([[2, 34.0, 0, 0, 13.0, 1, False, 0, 0, 1],
        [2, 31.0, 1, 1, 26.25, 0, False, 0, 0, 1],
        [1, 11.0, 1, 2, 120.0, 1, False, 0, 0, 1],
        [3, 0.42, 0, 1, 8.5167, 1, False, 1, 0, 0],
        [3, 27.0, 0, 0, 6.975, 1, False, 0, 0, 1],
        [3, 31.0, 0, 0, 7.775, 1, False, 0, 0, 1],
        [1, 39.0, 0, 0, 0.0, 1, False, 0, 0, 1],
        [3, 18.0, 0, 0, 7.775, 0, False, 0, 0, 1],
        [2, 39.0, 0, 0, 13.0, 1, False, 0, 0, 1],
        [1, 33.0, 1, 0, 53.1, 0, False, 0, 0, 1],
        [3, 26.0, 0, 0, 7.8875, 1, False, 0, 0, 1],
        [3, 39.0, 0, 0, 24.15, 1, False, 0, 0, 1],
        [2, 35.0, 0, 0, 10.5, 1, False, 0, 0, 1],
        [3, 6.0, 4, 2, 31.275, 0, False, 0, 0, 1],
        [3, 30.5, 0, 0, 8.05, 1, False, 0, 0, 1],
        [1, 29.69911764705882, 0, 0, 0.0, 1, True, 0, 0, 1],
        [3, 23.0, 0, 0, 7.925, 0, False, 0, 0, 1],
        [2, 31.0, 1, 1, 37.0042, 1, False, 1, 0, 0],
        [3, 43.0, 0, 0, 6.45, 1, False, 0, 0, 1],
        [3, 10.0, 3, 2, 27.9, 1, False, 0, 0, 1],
        [1, 52.0, 1, 1, 93.5, 0, False, 0, 0, 1],
        [3, 27.0, 0, 0, 8.6625, 1, False, 0, 0, 1],
        [1, 38.0, 0, 0, 0.0, 1, False, 0, 0, 1],
        [3, 27.0, 0, 1, 12.475, 0, False, 0, 0, 1],
        [3, 2.0, 4, 1, 39.6875, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 0, 0, 6.95, 1, True, 0, 1, 0],
        [3, 29.69911764705882, 0, 0, 56.4958, 1, True, 0, 0, 1],
        [2, 1.0, 0, 2, 37.0042, 1, False, 1, 0, 0],
        [3, 29.69911764705882, 0, 0, 7.75, 1, True, 0, 1, 0],
        [1, 62.0, 0, 0, 80.0, 0, False, 0, 0, 0],
        [3, 15.0, 1, 0, 14.4542, 0, False, 1, 0, 0],
        [2, 0.83, 1, 1, 18.75, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 0, 0, 7.2292, 1, True, 1, 0, 0],
        [3, 23.0, 0, 0, 7.8542, 1, False, 0, 0, 1],
        [3, 18.0, 0, 0, 8.3, 1, False, 0, 0, 1],
        [1, 39.0, 1, 1, 83.1583, 0, False, 1, 0, 0],
        [3, 21.0, 0, 0, 8.6625, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 0, 0, 8.05, 1, True, 0, 0, 1],
        [3, 32.0, 0, 0, 56.4958, 1, False, 0, 0, 1],
        [1, 29.69911764705882, 0, 0, 29.7, 1, True, 1, 0, 0],
        [3, 20.0, 0, 0, 7.925, 1, False, 0, 0, 1],
        [2, 16.0, 0, 0, 10.5, 1, False, 0, 0, 1],
        [1, 30.0, 0, 0, 31.0, 0, False, 1, 0, 0],
        [3, 34.5, 0, 0, 6.4375, 1, False, 1, 0, 0],
        [3, 17.0, 0, 0, 8.6625, 1, False, 0, 0, 1],
        [3, 42.0, 0, 0, 7.55, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 8, 2, 69.55, 1, True, 0, 0, 1],
        [3, 35.0, 0, 0, 7.8958, 1, False, 1, 0, 0],
        [2, 28.0, 0, 1, 33.0, 1, False, 0, 0, 1],
        [1, 29.69911764705882, 1, 0, 89.1042, 0, True, 1, 0, 0],
        [3, 4.0, 4, 2, 31.275, 1, False, 0, 0, 1],
        [3, 74.0, 0, 0, 7.775, 1, False, 0, 0, 1],
        [3, 9.0, 1, 1, 15.2458, 0, False, 1, 0, 0],
        [1, 16.0, 0, 1, 39.4, 0, False, 0, 0, 1],
        [2, 44.0, 1, 0, 26.0, 0, False, 0, 0, 1],
        [3, 18.0, 0, 1, 9.35, 0, False, 0, 0, 1],
        [1, 45.0, 1, 1, 164.8667, 0, False, 0, 0, 1],
        [1, 51.0, 0, 0, 26.55, 1, False, 0, 0, 1],
        [3, 24.0, 0, 3, 19.2583, 0, False, 1, 0, 0],
        [3, 29.69911764705882, 0, 0, 7.2292, 1, True, 1, 0, 0],
        [3, 41.0, 2, 0, 14.1083, 1, False, 0, 0, 1],
        [2, 21.0, 1, 0, 11.5, 1, False, 0, 0, 1],
        [1, 48.0, 0, 0, 25.9292, 0, False, 0, 0, 1],
        [3, 29.69911764705882, 8, 2, 69.55, 0, True, 0, 0, 1],
        [2, 24.0, 0, 0, 13.0, 1, False, 0, 0, 1],
        [2, 42.0, 0, 0, 13.0, 0, False, 0, 0, 1],
        [2, 27.0, 1, 0, 13.8583, 0, False, 1, 0, 0],
        [1, 31.0, 0, 0, 50.4958, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 0, 0, 9.5, 1, True, 0, 0, 1],
        [3, 4.0, 1, 1, 11.1333, 1, False, 0, 0, 1],
        [3, 26.0, 0, 0, 7.8958, 1, False, 0, 0, 1],
        [1, 47.0, 1, 1, 52.5542, 0, False, 0, 0, 1],
        [1, 33.0, 0, 0, 5.0, 1, False, 0, 0, 1],
        [3, 47.0, 0, 0, 9.0, 1, False, 0, 0, 1],
        [2, 28.0, 1, 0, 24.0, 0, False, 1, 0, 0],
        [3, 15.0, 0, 0, 7.225, 0, False, 1, 0, 0],
        [3, 20.0, 0, 0, 9.8458, 1, False, 0, 0, 1],
        [3, 19.0, 0, 0, 7.8958, 1, False, 0, 0, 1],
        [3, 29.69911764705882, 0, 0, 7.8958, 1, True, 0, 0, 1],
        [1, 56.0, 0, 1, 83.1583, 0, False, 1, 0, 0],
        [2, 25.0, 0, 1, 26.0, 0, False, 0, 0, 1],
        [3, 33.0, 0, 0, 7.8958, 1, False, 0, 0, 1],
        [3, 22.0, 0, 0, 10.5167, 0, False, 0, 0, 1],
        [2, 28.0, 0, 0, 10.5, 1, False, 0, 0, 1],
        [3, 25.0, 0, 0, 7.05, 1, False, 0, 0, 1],
        [3, 39.0, 0, 5, 29.125, 0, False, 0, 1, 0],
        [2, 27.0, 0, 0, 13.0, 1, False, 0, 0, 1],
        [1, 19.0, 0, 0, 30.0, 0, False, 0, 0, 1],
        [3, 29.69911764705882, 1, 2, 23.45, 0, True, 0, 0, 1],
        [1, 26.0, 0, 0, 30.0, 1, False, 1, 0, 0],
        [3, 32.0, 0, 0, 7.75, 1, False, 0, 1, 0]], dtype=object),
 'predictors': array([[3, 22.0, 1, ..., 0, 0, 1],
        [1, 38.0, 1, ..., 1, 0, 0],
        [3, 26.0, 0, ..., 0, 0, 1],
        ..., 
        [3, 31.0, 0, ..., 0, 0, 1],
        [3, 30.0, 0, ..., 1, 0, 0],
        [3, 30.0, 1, ..., 0, 0, 1]], dtype=object),
 'quit': <IPython.core.autocall.ExitAutocall at 0x7f55b790e400>,
 'target': array([[ 1.,  0.],
        [ 0.,  1.],
        [ 0.,  1.],
        ..., 
        [ 0.,  1.],
        [ 1.,  0.],
        [ 1.,  0.]]),
 'to_categorical': <function keras.utils.np_utils.to_categorical>,
 'train_size': 800}

In [20]:
# Specify, compile, and fit the model
model = Sequential()
model.add(Dense(32, activation='relu', input_shape = (n_cols,)))
model.add(Dense(2, activation='softmax'))
model.compile(optimizer='sgd', 
              loss='categorical_crossentropy', 
              metrics=['accuracy'])
model.fit(predictors, target)

# Calculate predictions: predictions
predictions = model.predict(pred_data)

# Calculate predicted probability of survival: predicted_prob_true
predicted_prob_true = predictions[:,1]

# print predicted_prob_true
print(predicted_prob_true)

Epoch 1/10
891/891 [==============================] - 0s 384us/step - loss: 2.4199 - acc: 0.6117
Epoch 2/10
891/891 [==============================] - 0s 46us/step - loss: 1.3396 - acc: 0.6263
Epoch 3/10
891/891 [==============================] - 0s 47us/step - loss: 0.8734 - acc: 0.6521
Epoch 4/10
891/891 [==============================] - 0s 46us/step - loss: 0.6267 - acc: 0.6846
Epoch 5/10
891/891 [==============================] - 0s 43us/step - loss: 0.6934 - acc: 0.6622
Epoch 6/10
891/891 [==============================] - 0s 44us/step - loss: 0.6203 - acc: 0.6880
Epoch 7/10
891/891 [==============================] - 0s 46us/step - loss: 0.6466 - acc: 0.7015
Epoch 8/10
891/891 [==============================] - 0s 45us/step - loss: 0.5927 - acc: 0.7160
Epoch 9/10
891/891 [==============================] - 0s 45us/step - loss: 0.6081 - acc: 0.7037
Epoch 10/10
891/891 [==============================] - 0s 46us/step - loss: 0.6047 - acc: 0.7003
[ 0.25012502  0.40163684  0.78297651  

Epoch 1/10

 32/800 [>.............................] - ETA: 0s - loss: 1.9889 - acc: 0.3438
672/800 [========================>.....] - ETA: 0s - loss: 2.1642 - acc: 0.5878
800/800 [==============================] - 0s - loss: 2.0926 - acc: 0.5850     
Epoch 2/10

 32/800 [>.............................] - ETA: 0s - loss: 0.9758 - acc: 0.7188
672/800 [========================>.....] - ETA: 0s - loss: 1.2621 - acc: 0.5789
800/800 [==============================] - 0s - loss: 1.2150 - acc: 0.5825     
Epoch 3/10

 32/800 [>.............................] - ETA: 0s - loss: 0.7413 - acc: 0.6562
704/800 [=========================>....] - ETA: 0s - loss: 0.9169 - acc: 0.6293
800/800 [==============================] - 0s - loss: 0.9013 - acc: 0.6225     
Epoch 4/10

 32/800 [>.............................] - ETA: 0s - loss: 0.8295 - acc: 0.3438
704/800 [=========================>....] - ETA: 0s - loss: 0.6976 - acc: 0.6335
800/800 [==============================] - 0s - loss: 0.7067 - acc: 0.6350     
Epoch 5/10

 32/800 [>.............................] - ETA: 0s - loss: 0.7734 - acc: 0.5312
672/800 [========================>.....] - ETA: 0s - loss: 0.7101 - acc: 0.6771
800/800 [==============================] - 0s - loss: 0.6920 - acc: 0.6775     
Epoch 6/10

 32/800 [>.............................] - ETA: 0s - loss: 0.7822 - acc: 0.4688
704/800 [=========================>....] - ETA: 0s - loss: 0.6394 - acc: 0.6705
800/800 [==============================] - 0s - loss: 0.6375 - acc: 0.6687     
Epoch 7/10

 32/800 [>.............................] - ETA: 0s - loss: 0.5666 - acc: 0.6250
640/800 [=======================>......] - ETA: 0s - loss: 0.6170 - acc: 0.6734
800/800 [==============================] - 0s - loss: 0.6172 - acc: 0.6812     
Epoch 8/10

 32/800 [>.............................] - ETA: 0s - loss: 0.6544 - acc: 0.6250
384/800 [=============>................] - ETA: 0s - loss: 0.6322 - acc: 0.6693
800/800 [==============================] - 0s - loss: 0.6181 - acc: 0.6837     
Epoch 9/10

 32/800 [>.............................] - ETA: 0s - loss: 0.7973 - acc: 0.5000
704/800 [=========================>....] - ETA: 0s - loss: 0.6325 - acc: 0.6634
800/800 [==============================] - 0s - loss: 0.6183 - acc: 0.6737     
Epoch 10/10

 32/800 [>.............................] - ETA: 0s - loss: 0.5755 - acc: 0.6875
672/800 [========================>.....] - ETA: 0s - loss: 0.6201 - acc: 0.6860
800/800 [==============================] - 0s - loss: 0.6142 - acc: 0.6900     
[ 0.18310049  0.39257559  0.90957892  0.43802381  0.19449668  0.16226816
  0.04776059  0.36320636  0.13548838  0.59659457  0.21850654  0.24186897
  0.14519456  0.54658085  0.1703662   0.09650292  0.3025246   0.50268072
  0.06854767  0.53118235  0.66011029  0.21718189  0.05112241  0.28982982
  0.58332694  0.18461925  0.58715367  0.68001962  0.19473749  0.59099054
  0.44369563  0.56261528  0.1525811   0.25684106  0.31957388  0.68460965
  0.2890453   0.19117369  0.59348351  0.42708611  0.29303789  0.36315283
  0.45844418  0.09916571  0.334088    0.07964922  0.56860393  0.10694957
  0.50475955  0.78340077  0.54485911  0.00884945  0.49051741  0.63139898
  0.25409591  0.3637715   0.93466604  0.15533349  0.3382667   0.1525811
  0.15455711  0.32354257  0.20677377  0.57712185  0.29516548  0.13900402
  0.28757289  0.58676606  0.21134241  0.50062877  0.21862327  0.52192533
  0.10416862  0.0638297   0.38311687  0.35487503  0.31126469  0.3045193
  0.18921892  0.61726612  0.47175369  0.14496818  0.34283844  0.21893513
  0.21895605  0.26417106  0.26272085  0.54475427  0.39032492  0.49731082
  0.15875942]